navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt

```diff
@@ -0,0 +1,129 @@
+}
+
+/**
+ * Navigator that drives a [ModalBottomSheetState] for use of [ModalBottomSheetLayout]s
+ * with the navigation library. Every destination using this Navigator must set a valid
+ * [Composable] by setting it directly on an instantiated [Destination] or calling
+ * [androidx.navigation.compose.material.bottomSheet].
+ *
+ * <b>The [sheetContent] [Composable] will always host the latest entry of the back stack. When
+ * navigating from a [BottomSheetNavigator.Destination] to another
+ * [BottomSheetNavigator.Destination], the content of the sheet will be replaced instead of a
+ * new bottom sheet being shown.</b>
+ *
+ * When the sheet is dismissed by the user, the [state]'s [NavigatorState.backStack] will be popped.
+ *
+ * @param sheetState The [ModalBottomSheetState] that the [BottomSheetNavigator] will use to
+ * drive the sheet state
+ */
+@OptIn(ExperimentalMaterialApi::class)
+@Navigator.Name("BottomSheetNavigator")
+public class BottomSheetNavigator(
+    val sheetState: ModalBottomSheetState
+) : Navigator<BottomSheetNavigator.Destination>() {
+
+    /**
+     * A [Composable] function that hosts the current sheet content. This should be set as
+     * sheetContent of your [ModalBottomSheetLayout].
+     */
+    public val sheetContent: @Composable ColumnScope.() -> Unit = @Composable {
+        val columnScope = this
+        val saveableStateHolder = rememberSaveableStateHolder()
+        val backStackEntries by state.backStack.collectAsState()
+        val transitionsInProgress by state.transitionsInProgress.collectAsState()
+
+        // We always replace the sheet's content instead of overlaying and nesting floating
+        // window destinations. That means that only *one* concurrent destination is supported by
+        // this navigator.
+        val latestEntry = backStackEntries.lastOrNull { entry ->
+            // We might have entries in the back stack that aren't started currently, so filter
+            // these
+            entry.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)
+        }
+
+        SheetContentHost(
+            columnHost = columnScope,
+            backStackEntry = latestEntry,
+            sheetState = sheetState,
+            saveableStateHolder = saveableStateHolder,
+            onSheetShown = { backStackEntry ->
+                val transitionForEntry = transitionsInProgress[backStackEntry]
+                transitionForEntry?.onTransitionComplete()
+            },
+            onSheetDismissed = { backStackEntry ->
+                state.pop(backStackEntry, saveState = false)
+            }
+        )
+    }
+
+    override fun createDestination(): Destination = Destination(navigator = this, content = {})
+
+    override fun navigate(
+        entries: List<NavBackStackEntry>,
+        navOptions: NavOptions?,
+        navigatorExtras: Extras?
+    ) {
+        entries.forEach { entry ->
+            state.pushWithTransition(entry)
```

Comment:
If we're `pushWithTransition` multiple entries (either because multiple entries are passed to this method or `navigate` is called multiple times in a row (which is probably the more likely case - that's how deep linking works), but we only display the latest one, how do those other entries get `onTransitionComplete` called?

Reply:
Done in e2d181a